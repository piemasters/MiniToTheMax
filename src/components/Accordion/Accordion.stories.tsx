import type { Meta, StoryObj } from '@storybook/react-vite';

import { Accordion, AccordionItem } from './Accordion';
import { BadgeNames } from '../../types';

const meta: Meta<typeof Accordion> = {
  component: Accordion,
  argTypes: {
    // type: {
    //   name: 'type',
    //   description: 'The badge type to display',
    //   defaultValue: BadgeNames.showcase,
    //   control: { type: 'radio' },
    // },
  },
};

export default meta;
type Story = StoryObj<typeof Accordion>;

export const Primary: Story = {
  render: () => {
    return (
      <div className="flex flex-col items-center justify-center min-h-screen">
        <Accordion className="max-w-lg">
          <AccordionItem value="1" trigger="ðŸ‘‹ Hello There">
            A demo Accordion item
          </AccordionItem>
          <AccordionItem value="2" trigger="ðŸŒŸ It's Animated">
            It transitions between the open and close states
          </AccordionItem>
          <AccordionItem value="3" trigger={<div>ðŸ§¶It's customisable</div>}>
            It is entirely customizable. You can put any HTML element and style
            it however you want.
          </AccordionItem>
          <AccordionItem value="4" trigger={<div>ðŸ¦• React & Tailwind</div>}>
            Nothing but React and Tailwind CSS
          </AccordionItem>
        </Accordion>
      </div>
    );
  },
  //   args: {
  //     type: BadgeNames.showcase,
  //   },
};
