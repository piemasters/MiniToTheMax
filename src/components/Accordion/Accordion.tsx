import {
  createContext,
  useContext,
  useRef,
  useEffect,
  useState,
  HTMLProps,
  ReactNode,
} from 'react';

const AccordionContext = createContext<
  | {
      selected: string | number | readonly string[] | undefined;
      setSelected: React.Dispatch<
        React.SetStateAction<string | number | readonly string[] | undefined>
      >;
    }
  | undefined
>(undefined);

type AccordionType = HTMLProps<HTMLUListElement> & {
  value?: string | number | readonly string[];
  onChange?: (value: string | number | readonly string[] | undefined) => void;
};

export const Accordion = ({
  children,
  value,
  onChange,
  ...props
}: AccordionType) => {
  const [selected, setSelected] = useState(value);

  useEffect(() => {
    onChange?.(selected);
  }, [selected]);

  return (
    <ul data-testid="accordion" {...props} className="m-0!">
      <AccordionContext.Provider value={{ selected, setSelected }}>
        {children}
      </AccordionContext.Provider>
    </ul>
  );
};

type AccordionItemType = HTMLProps<HTMLLIElement> & {
  trigger: ReactNode;
};

export const AccordionItem = ({
  children,
  value,
  trigger,
  ...props
}: AccordionItemType) => {
  const context = useContext(AccordionContext);
  if (!context) {
    throw new Error('AccordionItem must be used within an Accordion');
  }
  const { selected, setSelected } = context;
  const open = selected === value;

  const ref = useRef<HTMLDivElement>(null);

  return (
    <li
      data-testid="accordion-item"
      className="list-none! bg-white border-b"
      {...props}
    >
      <header
        role="button"
        tabIndex={!isNaN(Number(value)) ? Number(value) : 0}
        onClick={() => setSelected(open ? undefined : value)}
        onKeyDown={() => setSelected(open ? undefined : value)}
        className="flex items-center justify-between p-4 font-medium cursor-pointer bg-neutral-100 hover:bg-neutral-200"
      >
        <div
          data-testid="accordion-item-trigger"
          className="flex items-center justify-between p-4 font-medium"
        >
          {trigger}

          <ChevronDown
            className={`size-4 transition-transform ${open ? 'rotate-180' : ''}`}
          />
        </div>
      </header>
      <div
        data-testid="accordion-item-content"
        className={`overflow-y-hidden transition-all ${open ? 'open' : 'closed'}`}
        style={{ height: open ? ref.current?.offsetHeight || 0 : 0 }}
      >
        <div className="p-4" ref={ref}>
          {children}
        </div>
      </div>
    </li>
  );
};

type ChevronDownType = HTMLProps<HTMLDivElement> & {};

const ChevronDown = ({ className }: ChevronDownType) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      fill="none"
      viewBox="0 0 24 24"
      strokeWidth={1.5}
      stroke="currentColor"
      className={className}
    >
      <path
        strokeLinecap="round"
        strokeLinejoin="round"
        d="m19.5 8.25-7.5 7.5-7.5-7.5"
      />
    </svg>
  );
};
