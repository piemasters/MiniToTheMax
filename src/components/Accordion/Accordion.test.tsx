import {
  fireEvent,
  render,
  getByTestId,
  getAllByTestId,
} from '@testing-library/react';

import { Accordion, AccordionItem } from './Accordion';

describe('Accordion', () => {
  test('renders correctly', () => {
    const { container } = render(
      <Accordion className="max-w-lg">
        <AccordionItem value="1" trigger="ðŸ‘‹ Hello There">
          A demo Accordion item
        </AccordionItem>
        <AccordionItem value="2" trigger="ðŸŒŸ It's Animated">
          It transitions between the open and close states
        </AccordionItem>
        <AccordionItem value="3" trigger={<div>ðŸ§¶It's customisable</div>}>
          It is entirely customizable. You can put any HTML element and style it
          however you want.
        </AccordionItem>
        <AccordionItem value="4" trigger={<div>ðŸ¦• React & Tailwind</div>}>
          Nothing but React and Tailwind CSS
        </AccordionItem>
      </Accordion>
    );

    expect(getByTestId(container, 'accordion')).toBeVisible();
    expect(getAllByTestId(container, 'accordion-item').length).toBe(4);

    expect(getAllByTestId(container, 'accordion-item-content')[0]).toHaveClass(
      'closed'
    );
    fireEvent.click(getAllByTestId(container, 'accordion-item-trigger')[0]);
    expect(getAllByTestId(container, 'accordion-item-content')[0]).toHaveClass(
      'open'
    );

    expect(container).toMatchSnapshot();
  });
});
