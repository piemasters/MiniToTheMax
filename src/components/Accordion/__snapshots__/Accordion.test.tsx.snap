// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Accordion renders correctly 1`] = `
<div>
  <ul
    class="m-0!"
    data-testid="accordion"
  >
    <li
      class="list-none! bg-white border-b"
      data-testid="accordion-item"
    >
      <header
        class="flex items-center justify-between p-4 font-medium cursor-pointer bg-neutral-100 hover:bg-neutral-200"
        role="button"
        tabindex="1"
      >
        <div
          class="flex items-center justify-between p-4 font-medium"
          data-testid="accordion-item-trigger"
        >
          ðŸ‘‹ Hello There
          <svg
            class="size-4 transition-transform rotate-180"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m19.5 8.25-7.5 7.5-7.5-7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </header>
      <div
        class="overflow-y-hidden transition-all open"
        data-testid="accordion-item-content"
        style="height: 0px;"
      >
        <div
          class="p-4"
        >
          A demo Accordion item
        </div>
      </div>
    </li>
    <li
      class="list-none! bg-white border-b"
      data-testid="accordion-item"
    >
      <header
        class="flex items-center justify-between p-4 font-medium cursor-pointer bg-neutral-100 hover:bg-neutral-200"
        role="button"
        tabindex="2"
      >
        <div
          class="flex items-center justify-between p-4 font-medium"
          data-testid="accordion-item-trigger"
        >
          ðŸŒŸ It's Animated
          <svg
            class="size-4 transition-transform "
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m19.5 8.25-7.5 7.5-7.5-7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </header>
      <div
        class="overflow-y-hidden transition-all closed"
        data-testid="accordion-item-content"
        style="height: 0px;"
      >
        <div
          class="p-4"
        >
          It transitions between the open and close states
        </div>
      </div>
    </li>
    <li
      class="list-none! bg-white border-b"
      data-testid="accordion-item"
    >
      <header
        class="flex items-center justify-between p-4 font-medium cursor-pointer bg-neutral-100 hover:bg-neutral-200"
        role="button"
        tabindex="3"
      >
        <div
          class="flex items-center justify-between p-4 font-medium"
          data-testid="accordion-item-trigger"
        >
          <div>
            ðŸ§¶It's customisable
          </div>
          <svg
            class="size-4 transition-transform "
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m19.5 8.25-7.5 7.5-7.5-7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </header>
      <div
        class="overflow-y-hidden transition-all closed"
        data-testid="accordion-item-content"
        style="height: 0px;"
      >
        <div
          class="p-4"
        >
          It is entirely customizable. You can put any HTML element and style it however you want.
        </div>
      </div>
    </li>
    <li
      class="list-none! bg-white border-b"
      data-testid="accordion-item"
    >
      <header
        class="flex items-center justify-between p-4 font-medium cursor-pointer bg-neutral-100 hover:bg-neutral-200"
        role="button"
        tabindex="4"
      >
        <div
          class="flex items-center justify-between p-4 font-medium"
          data-testid="accordion-item-trigger"
        >
          <div>
            ðŸ¦• React & Tailwind
          </div>
          <svg
            class="size-4 transition-transform "
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m19.5 8.25-7.5 7.5-7.5-7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </header>
      <div
        class="overflow-y-hidden transition-all closed"
        data-testid="accordion-item-content"
        style="height: 0px;"
      >
        <div
          class="p-4"
        >
          Nothing but React and Tailwind CSS
        </div>
      </div>
    </li>
  </ul>
</div>
`;
